core/miner/storage_fsm/impl/commit_batcher_impl.cpp

```diff
@@ -0,0 +1,246 @@
+    infos.reserve(total);
+
+    BigInt collateral = 0;
+
+    for (const auto &[sector_number, pair_storage] : pair_storage_for_send) {
+      OUTCOME_TRY(sector_collateral,
+                  getSectorCollateral(sector_number, head->key));
+      collateral = collateral + sector_collateral;
+
+      params.sectors.insert(sector_number);
+      infos.push_back(pair_storage.aggregate_input.info);
+    }
+
+    for (const auto &[sector_number, pair_storage] : pair_storage_for_send) {
+      proofs.push_back(pair_storage.aggregate_input.proof);
+    }
+
+    const ActorId mid = miner_address_.getId();
+    // TODO maybe long (AggregateSealProofs)
+    AggregateSealVerifyProofAndInfos aggregate_seal{
+        .miner = mid,
+        .seal_proof =
+            pair_storage_for_send.at(infos[0].number).aggregate_input.spt,
+        .aggregate_proof = arp_,
+        .infos = infos};
+
+    std::vector<BytesIn> proofsSpan;
+    proofsSpan.reserve(proofs.size());
+
+    for (const Proof &proof : proofs) {
+      proofsSpan.push_back(gsl::make_span(proof));
+    }
+
+    OUTCOME_TRY(proof_->aggregateSealProofs(aggregate_seal, proofsSpan));
+
+    params.proof = aggregate_seal.proof;
+    OUTCOME_TRY(encode, codec::cbor::encode(params));
+    OUTCOME_TRY(miner_info, api_->StateMinerInfo(miner_address_, head->key));
+
+    const TokenAmount max_fee =
+        fee_config_->max_commit_batch_gas_fee.FeeForSector(proofs.size());
+
+    OUTCOME_TRY(tipset, api_->ChainGetTipSet(head->key));
+    const BigInt base_fee = tipset->blks[0].parent_base_fee;
+
+    TokenAmount agg_fee_raw = AggregateProveCommitNetworkFee(
+        infos.size(), base_fee);  // TODO change to aggregateNetworkFee
+
+    TokenAmount agg_fee = bigdiv(agg_fee_raw * agg_fee_num_, agg_fee_den_);
+    TokenAmount need_funds = collateral + agg_fee;
+    TokenAmount good_funds = max_fee + need_funds;
+
+    OUTCOME_TRY(address,
+                address_selector_(miner_info, good_funds, need_funds, api_));
+    OUTCOME_TRY(signed_messege,
+                api_->MpoolPushMessage(
+                    vm::message::UnsignedMessage(miner_address_,
+                                                 address,
+                                                 0,
+                                                 need_funds,
+                                                 max_fee,
+                                                 {},
+                                                 ProveCommitAggregate::Number,
+                                                 MethodParams{encode}),
+                    kPushNoSpec));
+
+    cutoff_start_ = std::chrono::system_clock::now();
+    return signed_messege.getCid();
+  }
+
+  void CommitBatcherImpl::setCommitCutoff(const ChainEpoch &current_epoch,
+                                          const SectorInfo &sector_info) {
+    ChainEpoch cutoff_epoch =
+        sector_info.ticket_epoch
+        + static_cast<int64_t>(kEpochsInDay + kChainFinality);
```

Comment:
it maybe `30*`.  
may move to miner policy header

https://github.com/filecoin-project/specs-actors/blob/v7.0.0/actors/builtin/miner/policy.go#L152-L164